### ISSUE TITLE
Allow linking tests on build

### ISSUE THREAD

**Issue by notYuriy** [ziglang/zig#7620]
To my knowledge, there is no way now to link tests in the final executable with `zig build` command. This feature can be potentially useful for implementing custom test runners. 

For example, if tests are written to run in an environment that is slightly different from the host one (e.g. unit tests for operating system kernel that require thread subsystem to be in place), they can't be simply executed with `zig test` command. 

As such, it would be useful to have an option to link all tests and make `builtin.test_functions` list accessible to a program if that is the case. Program will be able to set up the environment needed for tests to run and execute tests on its own.

**Comment by LemonBoy** [comment]
#6621 ?

**Comment by N00byEdge** [comment]
> #6621 ?
@LemonBoy 
I don't think so, this is about pulling the tests into the runtime, not the runtime into the tests.

**Comment by LemonBoy** [comment]
That's exactly what a custom test runner does.
You get a bunch of functions to run and you're free to run them however you want (sequentially or in parallel) and collect the results somehow.

**Comment by N00byEdge** [comment]
Are you sure? This is more of that you should be able to run the tests from the built binary, something like this:

``` 
pub main() void {
  for(@import("builtin").test_functions) |f|
    f.func();
}
```
or something similar, since the tests may depend on running in a different environment where the zig compiler isn't supported.

**Comment by LemonBoy** [comment]
`zig test` builds and runs an executable using [this code](https://github.com/ziglang/zig/blob/master/lib/std/special/test_runner.zig).

**Comment by N00byEdge** [comment]
Oh, I think I understand what you're saying. You're saying that with custom test runners you could build a different binary which just runs the tests instead? I suppose that would work too. If so, that could be the more general solution to this issue, but I think that either solution could be valid.

**Comment by andrewrk** [comment]
Duplicate of #6621

**Comment by notYuriy** [comment]
> `zig test` builds and runs an executable using [this code](https://github.com/ziglang/zig/blob/master/lib/std/special/test_runner.zig).
> Can we take control over the build process in this case? We want tests to be ran in emulator, and we want to only use zig build system for that

**Comment by PavelVozenilek** [comment]
One useful use case for such feature (custom test runner not limited to a special test mode): whenever program starts, it automatically executes tests from recently modified source files. This gives immediate feedback w/o the need to constantly think about running the tests.

**Comment by N00byEdge** [comment]
Yeah, in general I don't see a reason for `zig test` doing anything but providing `builtin.test_functions` to any environment (e.g. `zig test my_emulator_target`). That would mean the entire zig build system is available for tests, no custom test runner needed, your bootstrapping code can check using `@hasDecl(builtin, "test_functions")`, and decide what to do based on that.
